src/snapshots/overlay.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2021 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, Result};
+use nix::mount::MsFlags;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use crate::snapshots::{MountPoint, SnapShot, FS_OVERLAY};
+
+#[derive(Debug)]
+pub struct OverLay {
+    pub data_dir: PathBuf,
+    pub index: AtomicUsize,
+}
+
+impl SnapShot for OverLay {
+    fn mount(&mut self, layer_path: &[&str], mount_path: &Path) -> Result<MountPoint> {
+        let overlay_lowerdir = layer_path.join(":");
+        let index = self.index.fetch_add(1, Ordering::SeqCst).to_string();
+        let work_dir = self.data_dir.join(&index);
+        let overlay_upperdir = work_dir.join("upperdir");
+        let overlay_workdir = work_dir.join("workdir");
+
+        if !self.data_dir.exists() {
+            fs::create_dir_all(&self.data_dir)?;
+        }
+        fs::create_dir_all(&overlay_upperdir)?;
+        fs::create_dir_all(&overlay_workdir)?;
+
+        if !mount_path.exists() {
+            fs::create_dir_all(mount_path)?;
+        }
+
+        let source = Path::new(FS_OVERLAY);
+        let flags = MsFlags::empty();
+        let options = format!(
+            "lowerdir={},upperdir={},workdir={}",
+            overlay_lowerdir,
+            overlay_upperdir.display(),
+            overlay_workdir.display()
+        );
+
+        nix::mount::mount(
+            Some(source),
+            mount_path,
+            Some(FS_OVERLAY),
+            flags,
+            Some(options.as_str()),
+        )
+        .map_err(|e| {
+            anyhow!(
+                "failed to mount {:?} to {:?}, with error: {}",
+                source,
+                mount_path,
+                e
+            )
+        })?;
+
+        Ok(MountPoint {
+            r#type: FS_OVERLAY.to_string(),
+            mount_path: mount_path.to_path_buf(),
+            work_dir,
+        })
+    }
+
+    fn unmount(&self, mount_point: &MountPoint) -> Result<()> {
+        nix::mount::umount(mount_point.mount_path.as_path())?;
+        if mount_point.work_dir.exists() {
+            fs::remove_dir_all(&mount_point.work_dir)?;
+        }
+        Ok(())
+    }
```

Comment:
this function doesn't use `self` at all. I wonder if it makes sense to move it as a function on MountPoint itself.

Also I found no calls to `unmount()`, but I might missed something.

Reply:
@dubek Thanks for reviewing. Yes, currently we only implement the `pull_image` API which will setup the bundle, `remove_image` is not implemented yet, that API will call `unmount()` operation.

Since `SnapShot` trait need support diffent types which have different requirements for umount and clean operations,  for `OverLay` type, it is not needed. Also if this function can be shared by multiple snapshots, we can move it as function.

Reply:
Yes, I updated the `unmount` function and remove the `remove_dir_all` operations